apis/k8ssandra/v1alpha1/cassandraconfig_types.go

```diff
@@ -26,29 +26,29 @@ import (
 
 type CassandraConfig struct {
 	// +optional
-	CassandraYaml CassandraYaml `json:"cassandraYaml,omitempty" cass-config:"*:cassandra-yaml;recurse"`
+	CassandraYaml CassandraYaml `json:"cassandraYaml,omitempty" cass-config:"*:cassandra-yaml,dse:*:cassandra-yaml;recurse"`
 
 	// +optional
-	JvmOptions JvmOptions `json:"jvmOptions,omitempty" cass-config:"*:;recurse"`
+	JvmOptions JvmOptions `json:"jvmOptions,omitempty" cass-config:"*:,dse:*:;recurse"`
 
 	// +optional
-	DSEYaml DSEYaml `json:"dseYaml,omitempty" cass-config:"*:dse-yaml;recurse"`
+	DseYaml DseYaml `json:"dseYaml,omitempty" cass-config:"dse:*:dse-yaml;recurse"`
 }
 
 // CassandraYaml defines the contents of the cassandra.yaml file. For more info see:
 // https://cassandra.apache.org/doc/latest/cassandra/configuration/cass_yaml_file.html
 type CassandraYaml struct {
 	// Exists in 3.11, 4.0, trunk
 	// +optional
-	AllocateTokensForKeyspace *string `json:"allocate_tokens_for_keyspace,omitempty" cass-config:"*:allocate_tokens_for_keyspace"`
+	AllocateTokensForKeyspace *string `json:"allocate_tokens_for_keyspace,omitempty" cass-config:"*:allocate_tokens_for_keyspace,dse:*:allocate_tokens_for_keyspace"`
 
 	// Exists in: 4.0, trunk
 	// +optional
-	AllocateTokensForLocalReplicationFactor *int `json:"allocate_tokens_for_local_replication_factor,omitempty" cass-config:">=4.x:allocate_tokens_for_local_replication_factor"`
+	AllocateTokensForLocalReplicationFactor *int `json:"allocate_tokens_for_local_replication_factor,omitempty" cass-config:">=4.x:allocate_tokens_for_local_replication_factor,dse:>=6.8.x:allocate_tokens_for_local_replication_factor"`
```

Comment:
I used the following convention:

* For options available in all versions of Cassandra and DSE, use `dse:*`
* For options introduced in Cassandra 4.0 and DSE 6.8, use `dse:>=6.8.x`
* For options introduced in Cassandra 4.1, consider them absent from DSE 6.8.

Note that DSE 6.8 uses a fork of Cassandra 4.0 in reality. We would need to check each option one by one to make sure they are supported. We also might be missing a few ones present only in the 4.0 fork and not in the official 4.0. All of this could be made in a follow-up task.

Reply:
There's a huge difference in key counts in Configs, comparing 4.0.0, 4.0.6 and DSE. For example, 4.0.6 has these keys that 4.0.0 doesn't:

```
Keys in 4.0.6, not in 4.0.0: [otc_backlog_expiration_interval_ms cache_load_timeout_seconds allow_insecure_udfs force_new_prepared_statement_behaviour allow_extra_insecure_udfs]
```

4.0.6 has these keys that DSE doesn't have:

```
Keys in 4.0.6, not in DSE: [networking_cache_size_in_mb index_summary_capacity_in_mb internode_socket_receive_buffer_size_in_bytes max_streaming_retries autocompaction_on_startup_enabled internode_tcp_connect_timeout_in_ms enable_materialized_views keyspace_count_warn_threshold block_for_peers_in_remote_dcs automatic_sstable_upgrade internode_application_receive_queue_reserve_endpoint_capacity_in_bytes key_cache_migrate_during_compaction enable_transient_replication repaired_data_tracking_for_partition_reads_enabled concurrent_writes validation_preview_purge_head_start_in_sec concurrent_replicates concurrent_reads max_concurrent_automatic_sstable_upgrades credentials_cache_max_entries local_system_data_file_directory internode_socket_send_buffer_size_in_bytes check_for_duplicate_rows_during_reads cache_load_timeout_seconds snapshot_on_duplicate_row_detection repair_session_space_in_mb flush_compression native_transport_idle_timeout_in_ms internode_max_message_size_in_bytes repair_session_max_tree_depth creden
tials_validity_in_ms enable_sasi_indexes use_offheap_merkle_trees snapshot_links_per_second stream_entire_sstables diagnostic_events_enabled report_unconfirmed_repaired_data_mismatches credentials_update_interval_in_ms file_cache_enabled auto_optimise_preview_repair_streams replica_filtering_protection repaired_data_tracking_for_range_reads_enabled initial_range_tombstone_list_allocation_size table_count_warn_threshold internode_application_send_queue_capacity_in_bytes internode_tcp_user_timeout_in_ms native_transport_max_concurrent_requests_in_bytes enable_drop_compact_storage full_query_logging_options range_tombstone_list_growth_factor native_transport_max_concurrent_requests_in_bytes_per_ip auto_optimise_full_repair_streams force_new_prepared_statement_behaviour native_transport_max_negotiable_protocol_version consecutive_message_errors_threshold check_for_duplicate_rows_during_compaction internode_application_send_queue_reserve_endpoint_capacity_in_bytes snapshot_on_repaired_data_mismatch internode_strea
ming_tcp_user_timeout_in_ms corrupted_tombstone_strategy reject_repair_compaction_threshold native_transport_receive_queue_capacity_in_bytes block_for_peers_timeout_in_secs auto_optimise_inc_repair_streams concurrent_materialized_view_writes internode_application_receive_queue_capacity_in_bytes network_authorizer native_transport_flush_in_batches_legacy periodic_commitlog_sync_lag_block_in_ms index_summary_resize_interval_in_minutes concurrent_counter_writes internode_application_send_queue_reserve_global_capacity_in_bytes internode_application_receive_queue_reserve_global_capacity_in_bytes]
```

DSE has these keys that 4.0.0 doesn't have:

```
Keys in DSE, not in 4.0.0: [user_defined_function_fail_heap_mb nodesync user_defined_function_warn_micros backup_service seed_gossip_probability leaks_detection_params user_defined_function_warn_heap_mb tpc_io_cores native_transport_interface_prefer_ipv6 concurrent_lw_transactions max_memory_to_lock_mb encryption_options user_defined_function_fail_micros pick_level_on_streaming continuous_paging input_stream_channel_timeout_in_ms zerocopy_streaming_enabled allow_insecure_udfs sai_options permissions_cache_initial_capacity roles_cache_initial_capacity pick_level_on_streaming_min_sstable_size_bytes zerocopy_max_sstables io_global_queue_depth native_transport_address otc_small_connection_pool_size stream_outbound_permits_in_mb stream_outbound_buffer_in_kb snapshot_size_cache_validity_in_secs system_key_directory cross_dc_rtt_in_ms native_transport_interface otc_small_max_message_buffer_kb emulate_dbaas_defaults system_info_encryption batchlog_endpoint_strategy guardrails memtable_space_in_mb native_transport_bro
adcast_address otc_large_max_message_buffer_kb inflight_data_overhead_in_mb stream_max_outbound_buffers_in_kb internode_recv_buff_size_in_bytes native_transport_keepalive full_query_log_dir metadata_directory allow_extra_insecure_udfs internode_send_buff_size_in_bytes system_keyspaces_filtering echo_attempts_before_reset tpc_cores direct_reads_size_in_mb otc_backlog_expiration_interval_ms metrics_histogram_update_interval_millis snapshot_before_dropping_column max_memory_to_lock_fraction aggregated_request_timeout_in_ms tpc_pending_requests_limit tpc_concurrent_requests_limit zerocopy_max_unused_metadata_in_mb max_pending_lw_transactions]
```

Reply:
These can all be quite easily autogenerated, I used the code in cass-operator PR 224 to create those differences.

Reply:
> There's a huge difference in key counts in Configs, comparing 4.0.0, 4.0.6 and DSE. 

Yes, I am assuming _for now_ that DSE 6.8 = C* 4.0, but we all know that it's not true. we will have to revisit `CassandraYaml`. I don't think we need to do this now because imo it's not urgent.

> These can all be quite easily autogenerated

The problems with autogeneration for config options imo are:

1. These definitions are not a primary source of truth. There might be mistakes in the current definitions and we don't know. Actually I think I found one such mistake: there is a property called `index_options` declared in dse-yaml-dse-6.8.0.edn that is not recognized by DSE 6.8.19 (DSE could not start).
2. We lose the docs. I think docs are important, because otherwise users need to know where to find these docs.
3. If we get rid of config-builder completely, would we still generate this code from its definitions?

Reply:
What we could do is mix-and-match with EDN files. So read all the properties from the Java, but add descriptions from EDN files if there's any. That way new fields would be detected.